ddtrace/tracer/rules_sampler.go

```diff
@@ -54,6 +54,55 @@ func (r *rulesSampler) HasSpanRules() bool { return r.spans.enabled() }
 
 func (r *rulesSampler) TraceRateLimit() (float64, bool) { return r.traces.limit() }
 
+type Provenance int32
+
+const (
+	Local    Provenance = iota
+	Customer Provenance = 1
+	Dynamic  Provenance = 2
+)
+
+var (
+	ProvenanceName = map[Provenance]string{
+		Local:    "local",
+		Customer: "customer",
+		Dynamic:  "dynamic",
+	}
+	ProvenanceValue = map[string]Provenance{
+		"local":    Local,
+		"customer": Customer,
+		"dynamic":  Dynamic,
+	}
+)
+
+func (p Provenance) String() string {
+	return ProvenanceName[p]
+}
+
+func (p Provenance) MarshalJSON() ([]byte, error) {
+	return json.Marshal(p.String())
+}
+
+func (p *Provenance) UnmarshalJSON(data []byte) error {
```

Comment:
Also, would be nice to make the receivers of just pointer type.

Reply:
Custom MarshalJSON cannot have pointer receiver, otherwise, it won't get called when serializing an object, but not a pointer. On the other hand, an object receiver will make sure custom json marshal is called in both objects and receivers.

This is unfortunately some idiosyncracy.